lib/Server.js

```diff
@@ -732,23 +734,54 @@ class Server {
   listen(port, hostname, fn) {
     this.hostname = hostname;
 
-    return this.listeningApp.listen(port, hostname, (err) => {
+    const setupCallback = () => {
       this.createSocketServer();
 
       if (this.options.bonjour) {
         runBonjour(this.options);
       }
 
       this.showStatus();
+    };
 
-      if (fn) {
+    // between setupCallback and userCallback should be any other needed handling,
+    // specifically so that things are done in the right order to prevent
+    // backwards compatability issues
+    let userCallbackCalled = false;
+    const userCallback = (err) => {
+      if (fn && !userCallbackCalled) {
+        userCallbackCalled = true;
         fn.call(this.listeningApp, err);
       }
+    };
 
+    const onListeningCallback = () => {
       if (typeof this.options.onListening === 'function') {
         this.options.onListening(this);
       }
-    });
+    };
+
+    const fullCallback = (err) => {
+      setupCallback();
+      userCallback(err);
+      onListeningCallback();
+    };
```

Comment:
I think we should refactor this in future, looks very misleading

Reply:
Really those 3 functions can all be in `fullCallback`, since they are only called by `fullCallback`. Originally I was doing some other things with these individual functions to avoid a few breaking changes and stay organized, but do you think just putting all their functionality in that one callback will be cleaner?

Reply:
@Loonride i think we should allow `userCallback` to be async, also ehre good questions should `userCallback` will be called before or after `onListeningCallback`, potentially user callback should be called after all operation, it is allow do something with server (or not with server) when webpack-dev-server starts, for example for other web server, (for example if you use php project, you should use this callback to run php dev server)

Reply:
> i think we should allow userCallback to be async

Maybe we should do this in a separate PR?

> potentially user callback should be called after all operation

So this is a good order the way I have above, because `userCallback` is after `setupCallback`, right?

Reply:
> Maybe we should do this in a separate PR?

Yes, just thoughts

> So this is a good order the way I have above, because userCallback is after setupCallback, right?

oh, it is interesting, i think `userCallback` should be after all callbacks, because i want to get socket information after starts server

Reply:
> oh, it is interesting, i think userCallback should be after all callbacks, because i want to get socket information after starts server

With the current ordering, `userCallback` should be able to access socket information (because `this.createSocketServer()` is in `setupCallback`.)

And I think the ordering of these two:
```
userCallback(err);
onListeningCallback();
```

should not matter, as user should probably only use one or the other (the `userCallback` calls the callback for the `Server.listen` method, and the `onListeningCallback` calls the user provided `options.onListening` callback).

There is, however, a question of whether these two callbacks should always be called in the same places. For consistency, I would say yes. However, `onListening` currently gets called in these two places:

https://github.com/webpack/webpack-dev-server/pull/2061/files#diff-2e3826007762dc4188ae8303ccd8b036R25

https://github.com/webpack/webpack-dev-server/pull/2061/files#diff-2e3826007762dc4188ae8303ccd8b036R25

But these places are technically not the server listening, they are just errors. Could we possibly change the `onListening` option to accept both an error object and the server object? Optimally, it would be `onListening(err, server)`, but since the current api is `onListening(server)`, `onListening(server, err)` would reduce breaking changes. Thoughts?

Reply:
/cc @hiroppy what do you think about this too

Reply:
I want to change `onListening(err, server)`. In fact, this format is correct as Node.js theory.

Reply:
:+1: agree

Reply:
So can we get this merged, and then I'll do 2 PRs:

1. async `userCallback`
2. `onListening(err, server)`

But @evilebottnawi can you clarify the async one. A user can already do this without problems:

```
server.listen(port, host, async (err) => {
  await something();
  // ...
});
```

And I think there are no things before `userCallback` that are asynchronous. Maybe you mean we should do:

```
setupCallback();
await userCallback(err);
onListeningCallback();
```

But the only thing this would do is delay `onListeningCallback` until a potentially async `userCallback` is complete. 

Is this what you meant?

Reply:
> But @evilebottnawi can you clarify the async one? A user can already do this without problems:
```
server.listen(port, host, async (err) => {
  await something();
  // ...
});
```

Yep, but it is require to create own util for running webpack-dev-server, in some cases it is unnecessary, because you want run one simple task when  webpack-dev-server started

> But the only thing this would do is delay onListeningCallback until a potentially async userCallback is complete.

> Is this what you meant?

It is expected and can be used for some features